Martink rsa/heatmap

martink-rsa opened the pull request:
I've fulfilled all AC. 

1. Added a heatmap. A 2d array of numbers is passed to the heatmap to populate the data.
2. Added custom styling to each heatmap block based on the value.

Notes:
1. I didn't find handling the data too challenging as day and hour indexes are already used in JS functions. I started with the design first using dummy data, which took most of the time, then worked on parsing the data which was just a matter of plug-'n-play when finished. 
2. Styling has been the real challenge. The heatmap is an image in Zeplin so there are no values one can get, they will have to use their own intuition and experimentation. I feel a bit of further fine tuning and I'll have it mostly the same.
3. I'm unsure of how to stye the onHover for the heatmap blocks and have used opacity with a white background to simulate it lightening.

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

jkettmann commented on src/dummyData/heatmapDummyData.js:
```diff
@@ -0,0 +1,186 @@
+const heatmapDummyData = [
```
This file should be removed

jkettmann commented on src/pages/SearchPage/Heatmap/Heatmap.js:
```diff
@@ -0,0 +1,105 @@
+  'Saturday',
+];
+
+const heatmapColors = [
```
I think it's nicer to remove these colors and keep the ones in the theme

jkettmann commented on src/pages/SearchPage/Heatmap/Heatmap.js:
```diff
@@ -0,0 +1,105 @@
+              <InfoBlock
+                onClick={() => setSelected({ day: dayIndex, hour: hourIndex })}
+                enabled={selected.day === dayIndex && selected.hour === hourIndex}
+                background={heatmapColors[value]}
```
This is not working for values > 10. I would just pass the value to the InfoBlock component and handle the mapping of value to color inside a function inside the style file

jkettmann commented on src/pages/SearchPage/Heatmap/Heatmap.js:
```diff
@@ -0,0 +1,105 @@
+function Heatmap({ info }) {
+  const [selected, setSelected] = useState({});
+
+  return (
```
It's a bit of premature optimization, but if you want to show some performance optimization skills you could extract the `HoursContainer`, `DaysContainer` and `InfoContainer` into separate components. Then you could use [React.memo](https://reactjs.org/docs/react-api.html#reactmemo) on them to avoid unnecessary re-rendering.

You could further extract another component out of the InfoContainer that renders a row. This could use `React.memo` again. Now when the heatmap is re-rendered for example because an InfoBlock was selected only the one or two rows that were affected by the state change will be re-rendered (the row that includes the previously selected InfoBlock and the newly selected one)

jkettmann replied:
As I said this is premature optimization. But it might make a good impression on an employer

jkettmann commented on src/pages/SearchPage/Heatmap/Heatmap.js:
```diff
@@ -0,0 +1,105 @@
+          </InfoContainer>
+        </BottomRow>
+      </HeatmapContainer>
+      <TimeMessage>
```
This could be extracted into a separate component just to this file and the styles file a bit smaller

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -17,25 +17,53 @@ async function handleSearch(subreddit) {
+    const results = searchResults.data;
+    for (let i = 0; i < results.length; i += 1) {
+      // Add utc seconds from API to base epoch time
+      const date = new Date(0);
```
IMO this would look a bit simpler. But I'll leave it up to you.

```
const createdAtInMs = results[i].created_utc * 1000;
const date = new Date(createdAtInMs)
```


Would make the comment above unnecessary, and the less comments required the better. After all, you need to maintain comments as well. Meaning when the impementation changes the comment needs to be adjusted as well. When you forget that, you have a comment that points you in the wrong direction. And that's very confusing.

Sorry for the detour, just wanted to share that :-)

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -17,25 +17,53 @@ async function handleSearch(subreddit) {
+
   // Search for new subreddit when parameters have changed
   useEffect(() => {
+    setDataLoaded(false);
```
This flag is used to distinguish between "everything is ok" and "an error occurred", right? I'd rather use a `setError` state here, since you need to show the error anyways. You could then simply save a string to the error state and render that instead of the spinner and heatmap

martink-rsa replied:
I want to implement this but I'm not sure exactly what you mean. If it's not too much trouble, would you be able to expand on what you mean by saying to `save a string to the error state and render that instead of the spinner and heatmap`?

Thank you

jkettmann replied:
Sure. I mean instead of

```
const [dataLoaded, setDataLoaded] = useState(false);
setDataLoaded(false);
```

You could

```
const [error, setError] = useState(null);
setError(error.message);
// or simply
setError('An error occurred. Please try again')
```

Then you can also display the error instead of the heatmap like this or so

```
error && <div>{error}</div>
```

jkettmann commented on src/styles/theme.js:
```diff
@@ -42,6 +42,24 @@ const theme = {
   breakpoint: {
     small: '768px',
   },
+  heatmapColors: {
```
You have heatmapStyle already below. I'd move these colors there as well

jkettmann commented on src/docs/Heatmap.md:
```diff
@@ -0,0 +1,17 @@
+# Heatmap design docs
```
Well done doing this without the designs! [But they are actually there](https://app.zeplin.io/project/5e40457df723790a43273103/screen/5e4cfbc201c6a2a9a94587a3) and not just an image. Sorry for the additional work. In such a case you can ask the designer (in this case me) if there has been a mistake. Everything should be a proper design.

![image](https://user-images.githubusercontent.com/9722792/78536717-93929380-77ee-11ea-8309-ecbe27b1c670.png)

jkettmann commented on src/pages/SearchPage/Heatmap/Heatmap.style.js:
```diff
@@ -0,0 +1,126 @@
+  width: 960px;
+`;
+
+const InfoBlock = styled.button`
```
I think it would be good to remove the outline here so the element doesn't seem to "glow" when focused

jkettmann reviewed (changes requested):
About your comments:

1. Using dummy data is always a good idea. I think you might need to move away from using numbers in your array for the next task, but you can adjust it there.
2. Sorry for the additional work. The designs for the heatmap are actually there and not just an image. Did you only look at the home page? There's more info in one of the comments 
3. I guess the acceptance criteria are a bit unclear. You can use the same black border for hovering as we use for the selected state. No need to change the background color

Additionally a few things are slightly off. Especially the subreddit form and the timezone

![image](https://user-images.githubusercontent.com/9722792/78537347-a78ac500-77ef-11ea-9e87-b65343bf0375.png)

Btw nice job with the data handling. I guess in this case I overcomplicated things :-D

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa commented:
I've made all changes. ~~with the exception of the `setError` state.~~

1. I've extracted 4 different components as per your suggestion, as well as added `react.memo()` to them:
1.1.  Days, Hours and Info extracted to their own component.
1.2. InfoRow displays a row at a time.

2. I've gone grid-based layout to flex rows so that I can render a row at a time. I'm not sure if I've implemented the rows being rendered separately correctly. 

From my understanding, I would need to render out the row components separately and not in a loop, since React won't be able to access the individual row if it was generated from a loop? 

That was my line of thinking so I manually created a separate row for each day instead of mapping it out. 

Please could you advise on this?

3. I'm still struggling with a background image and the Perfect Pixel extension, and can't seem to get a reliable fit exactly each time. I can photoshop a perfect size from a screenshot I took of Zeplin, but it still takes adjusting in Perfect Pixel and playing around with the browser width. Here is how it looks currently on my screen for comparison, I'm not sure if I've managed to replicate it exactly:
![image](https://user-images.githubusercontent.com/52622303/78592432-5e656000-7845-11ea-93e8-59a8e505520a.png)

Edit:
I forgot to mention that I changed the way I handle data which now contains objects with info related to the Reddit post

4. I've changed the state to `error`, but I could only get the correct rendering sequence by have a double logical operator check: ` {(!isLoading && !error) && <Heatmap info={heatmapInfo} />}`

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

jkettmann commented on src/pages/SearchPage/Heatmap/Info/Info.style.js:
```diff
@@ -0,0 +1,36 @@
+  width: 960px;
+`;
+
+function getInfoBlockIndex(index) {
```
Maybe rename this to `getColorIndex`

jkettmann commented on src/pages/SearchPage/Heatmap/InfoRow/InfoRow.js:
```diff
@@ -0,0 +1,36 @@
+          onClick={() => setSelected({ day, hour: hourIndex })}
+          enabled={selected.day === day && selected.hour === hourIndex}
+          bgIndex={obj.dayHourCount}
+          key={`ib-${hourIndex.toString()}}`}
```
I guess `toString` is not required here!? `hourIndex` is just a number, right?

martink-rsa replied:
This is what React docs did when using an index. If I remove the `toString()`, ESLint has an error. Not sure what key to give it.

jkettmann replied:
You mean this warning? True. You could also disable the ESLint warning. But that has the same effect. The warning doesn't matter here because we don't change the order of the items.

![Screenshot 2020-04-07 16:57:58](https://user-images.githubusercontent.com/9722792/78684714-1865d600-78f1-11ea-8618-2d6a7c098f16.png)

 If you don't know what this warning means [have a look here](https://github.com/yannickcr/eslint-plugin-react/tree/master/docs/rules/no-array-index-key.md) btw. That is important when you render an array and change the order in some way.

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -4,66 +4,77 @@ import SearchPageWrapper from './SearchPage.style';
 
+function parseRedditData(input) {
+  const results = [...input.data];
+  const arrInfo = Array(7).fill().map(() => Array(24).fill().map(() => ({
```
You can simplify this by just using an array of posts like this

```
const arrInfo = Array(7).fill().map(() => Array(24).fill().map(() => [])
```

Now you can get the `dayHourCount` in the Info component via `Array.length`

jkettmann commented on src/pages/SearchPage/Heatmap/Heatmap.style.js:
```diff
@@ -23,104 +23,16 @@ const TopRow = styled.div`
 const BottomRow = styled.div`
   display: flex;
   flex-direction: rows;
   height: 280px;
```
You don't need this height since the height of 40px is already defined for the Info components

jkettmann commented on src/pages/SearchPage/Heatmap/Hours/Hours.style.js:
```diff
@@ -0,0 +1,35 @@
+  justify-content: center;
+  align-items: center;
+  height: 53px;
+  width: 80px;
```
You could extract the Info width into the theme and then calculate this width to 2*boxWidth

jkettmann commented on src/pages/SearchPage/Heatmap/Hours/Hours.style.js:
```diff
@@ -0,0 +1,35 @@
+  height: 53px;
+  width: 80px;
+  background: linear-gradient(180deg, #fefefe 0%, #e9e9e9 100%);
+  border-top: 1px solid #f3f3f3;
```
This color might go to the theme, not sure though

jkettmann commented on src/pages/SearchPage/Heatmap/Info/Info.js:
```diff
@@ -0,0 +1,32 @@
+
+  return (
+    <InfoContainer>
+      <InfoRow selected={selected} setSelected={setSelected} day={0} hours={info[0]} />
```
You can use an array here. Actually `info` is already an array right? Can't you just do `info.map(...` here?

jkettmann commented on src/pages/SearchPage/Heatmap/Info/Info.style.js:
```diff
@@ -0,0 +1,36 @@
+const InfoContainer = styled.div`
+  display: flex;
+  flex-direction: column;
+  width: 960px;
```
This width is not required

jkettmann commented on src/pages/SearchPage/Heatmap/InfoRow/InfoRow.style.js:
```diff
@@ -0,0 +1,39 @@
+
+const InfoRowWrapper = styled.div`
+  display: flex;
+  width: 960px;
```
Not required

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -4,66 +4,77 @@ import SearchPageWrapper from './SearchPage.style';
+    dayHourCount: 0,
+    posts: [],
+  })));
+  for (let i = 0; i < results.length; i += 1) {
```
You might be able to use `results.forEach`.

jkettmann commented on src/pages/SearchPage/Heatmap/Info/Info.style.js:
```diff
@@ -0,0 +1,36 @@
+  font-weight: 600;
+  line-height: 0.64;
+  &:hover {
+    border: solid 1px #1e2537;
```
This color is reused, so it can go to the theme

jkettmann reviewed (changes requested):
1. The usage of React.memo is great. The component structure looks also fine

2. The grid based layout was actually very nice. I think you still could have used it instead of flex. For example the InfoRow could look like this

```
function InfoRow({...}) {
  return hours.map(...)
}
```

So I guess the grid approach would work if the Info components wouldn't be wrapped in another element like

```
function InfoRow({...}) {
  return (
    <InfoRowWrapper>
      { hours.map(...) }
    <InfoRowWrapper>
}
```

If you like and still have time to change it feel free to try. Just be aware of the [browser limitations of display: grid](https://caniuse.com/#search=display%20grid)

3. The problem with the pixel-perfect is a bit weird. I also cannot find a perfect solution. One thing is that the font-size of the weekdays is 14px instead of 15px. Then the letter-spacing: 0.8px is not required anymore. But the heatmap seems to be off by 1px vertically. The timezone is off by 1px horizontally. I'll send you a screencast on slack (very boring though)

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa committed:
Heatmap

martink-rsa commented:
I've made all of the suggested changes with the exception of changing from Flexbox back to Grid.

1. I've used the screenshot you sent me and I believe I've got it 100% except for the logo which is a slightly different size on my version. Glad I'm finally able to follow something and know it will be correct at the end if I get it aligned correctly. Please let me know if anything is not aligned properly.

2. I've decided to stick with the flex containers for the rows for now. I am still thinking about switching to Grid or staying with Flex based on what will be better for getting the site to be mobile responsive.

martink-rsa committed:
Heatmap

jkettmann reviewed (approved):
Great work! It doesn't really matter whether you go with the grid or flex approach. You can always change later
